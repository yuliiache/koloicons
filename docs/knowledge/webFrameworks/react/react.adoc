== React
:toc:

=== 1. https://legacy.reactjs.org/docs/components-and-props.html#function-and-class-components[React component definition (functional, class)]
=== 2. https://react.dev/learn/passing-props-to-a-component[React component props]
- https://legacy.reactjs.org/docs/render-props.html[render props]

=== 3. https://react.dev/learn/state-a-components-memory[React component state]
- initial state
- setState
- stateful vs stateless components

=== 4. React class component lifecycle methods
=== 5.* https://react.dev/reference/react[React hooks API]:
- `useEffect (https://react.dev/learn/removing-effect-dependencies[problem])`, `useState`, `useReducer`, `useMemo`, `useRef`
- https://reactjs.org/docs/hooks-rules.html[rules]
- https://react.dev/learn/reusing-logic-with-custom-hooks[custom hooks]

=== 6. Components composition
- props.children
- composition vs inheritance
- lifting state up
- practical decomposition task

=== 7. https://react.dev/learn/writing-markup-with-jsx[JSX]
=== 8. https://react.dev/reference/react-dom[React DOM]
=== 9. https://react.dev/reference/react-dom/components/common#reference[React DOM Events]
`SyntheticEvent`, `persist event`

=== 10. https://react.dev/reference/react-dom/components#form-components:~:text=and%20dangerouslySetInnerHTML.-,Form%20components,-These%20built%2Din[Forms]
`controlled`,`uncontrolled components`

=== 11. https://legacy.reactjs.org/docs/faq-internals.html#what-is-the-virtual-dom[Virtual DOM]
- https://ru.reactjs.org/docs/reconciliation.html[reconciliation]
- https://github.com/acdlite/react-fiber-architecture[fiber]

=== 12. React code reuse patterns
- https://legacy.reactjs.org/docs/higher-order-components.html[higher-order components]

=== 13. https://react.dev/reference/react/createContext[Context]
`when to use`, `Context.Provider`, `Context.Consumer`, `Context vs Redux`

=== 14. https://react.dev/reference/react-dom/createPortal[Portals]
=== 15.  https://react.dev/reference/react/Component#catching-rendering-errors-with-an-error-boundary[Error Boundaries]

=== 16.  React component lifecycle phases (render, pre-commit, commit)
- https://vmarchesin.medium.com/the-react-lifecycle-step-by-step-47c0db0bfe73#:%7E:text=Pre%2Dcommit%20phase%3A%20Is%20a,any%20side%20effects%20are%20triggered[class components]
- https://react.dev/learn/render-and-commit[fn components]
- https://blog.isquaredsoftware.com/2020/05/blogged-answers-a-mostly-complete-guide-to-react-rendering-behavior/#render-and-commit-phases[...]

=== 19. https://react.dev/reference/react-dom/server[Server rendering]
`renderToPipeableStream`
`renderToStaticNodeStream`
`renderToReadableStream`
`renderToString`
`renderToNodeStream`


=== 20. https://react.dev/reference/react/memo[React.memo], https://react.dev/reference/react/lazy[React.lazy, Suspense p1] https://betterprogramming.pub/routing-and-code-splitting-in-react-fc7315f0bde4[p2]

=== 21. Optimizing Performance
- https://react.dev/learn/rendering-lists#keeping-list-items-in-order-with-key[keys]
- https://react.dev/reference/react/Profiler[Profiler],
- https://chrome.google.com/webstore/detail/react-developer-tools/fmkadmapgofadopljbjfkapdkoienihi?hl=en[Dev Tools],
- https://create-react-app.dev/docs/production-build/[prod build, code splitting],
- https://levelup.gitconnected.com/how-to-render-your-lists-faster-with-react-virtualization-5e327588c910[virtualization]

=== 22. Security
-  https://medium.com/javascript-security/avoiding-xss-in-react-is-still-hard-d2b5c7ad9412[XSS on React Props]
- https://blog.logrocket.com/using-dangerouslysetinnerhtml-in-a-react-application/[dangerouslySetInnerHTML]







