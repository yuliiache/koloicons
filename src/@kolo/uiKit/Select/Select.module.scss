@import "@kolo/design";

.select {
  position: relative;

  display: flex;
  flex-wrap: wrap;
  padding-bottom: 16px;

  &Label {
    display: block;
    margin-bottom: 8px;

    font-weight: 500;
    font-size: 12px;
    line-height: 1.33;
    color: $secFontColor;

    cursor: pointer;
  }

  &Wrapper {
    position: relative;

    display: block;
    width: 100%;
    min-width: 200px;
    height: 56px;
  }

  &Options {
    display: block;
    box-sizing: border-box;
    min-width: 100%;
    height: 100%;
    padding: 16px 16px;

    font-size: 0.8rem;
    line-height: 1.5;
    color: $secFontColor;

    background: linear-gradient(to bottom, $WhiteBG, $WhiteBG);

    border: solid 1px $ActiveBorderColor;
    border-radius: 4px;
    box-shadow: inset 0 2px 0 0 $WhiteBG, inset 0 24px 35px -4px $ShadowColor;
    cursor: pointer;
    appearance: none;

    transition: box-shadow 0.3s ease-in-out;

    &:hover,
    &:focus {
      background: linear-gradient(to bottom, $WhiteBG, $WhiteBG);
      outline: none;
      box-shadow: inset 0 2px 0 0 $WhiteBG, inset 0 24px 35px -4px $ActiveInputColor;

      & + svg {
        fill: $IconColor;
      }
    }

    &:focus {
      box-shadow: inset 0 2px 0 0 $WhiteBG, inset 0 24px 35px -4px $ActiveInputColor, 0 0 0 4px $FocusShadowInput;
    }

    &:invalid {
      color: $GreyColor;
    }

    &Option {
      color: $secFontColor;
    }

    & + svg {
      position: absolute;
      top: 50%;
      right: 16px;

      width: 16px;
      height: 16px;

      fill: $IconDisable;

      transform: translateY(-50%);
      transition: fill 0.3s ease-in-out;
    }
  }
}
