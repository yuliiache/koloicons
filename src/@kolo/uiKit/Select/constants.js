export const SELECT_TEST_ID = 'selectComponent';
