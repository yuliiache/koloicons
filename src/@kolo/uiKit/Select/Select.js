import React from 'react';
import PropTypes from 'prop-types';

import {SELECT_TEST_ID} from './constants';

import styles from './Select.module.scss';

export const Select = ({label, name, options = [], value = '', compId = 'selectId', onChange}) => {
  return (
    <div className={styles.select}>
      {!!label && (
        <label
          className={styles.selectLabel}
          data-testid={SELECT_TEST_ID}
          htmlFor={compId}
        >
          {label}
        </label>
      )}
      <div className={styles.selectWrapper}>
        <select
          className={styles.selectOptions}
          onChange={onChange}
          name={name}
          id={compId}
          value={value}
        >
          {options.length &&
            options.map((option) => (
              <option
                key={option}
                data-testid={option}
                className={styles.selectOption}
                value={option}
              >
                {option}
              </option>
            ))}
        </select>
      </div>
    </div>
  );
};

Select.propTypes = {
  label: PropTypes.string,
  name: PropTypes.string,
  options: PropTypes.arrayOf(PropTypes.string),
  value: PropTypes.string,
  compId: PropTypes.string,
  onChange: PropTypes.func,
};
