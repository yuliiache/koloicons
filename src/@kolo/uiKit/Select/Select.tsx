import {FC} from 'react';

import {SELECT_TEST_ID} from './constants';
import styles from './Select.module.scss';

interface Props {
  label?: string;
  name?: string;
  options?: string[];
  value?: string;
  compId?: string;
  onChange?: () => void;
}

export const Select: FC<Props> = ({label, name, options = [], value = '', compId = 'selectId', onChange}) => {
  return (
    <div className={styles.select}>
      {!!label && (
        <label
          className={styles.selectLabel}
          data-testid={SELECT_TEST_ID}
          htmlFor={compId}
        >
          {label}
        </label>
      )}
      <div className={styles.selectWrapper}>
        <select
          className={styles.selectOptions}
          onChange={onChange}
          name={name}
          id={compId}
          value={value}
        >
          {options.length &&
            options.map((option) => (
              <option
                key={option}
                data-testid={option}
                className={styles.selectOption}
                value={option}
              >
                {option}
              </option>
            ))}
        </select>
      </div>
    </div>
  );
};
