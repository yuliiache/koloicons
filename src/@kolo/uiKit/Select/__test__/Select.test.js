import React from 'react';
import {cleanup, render, screen} from '@testing-library/react';

import {Select} from '../Select';

const selectOptions = ['first', 'second', 'third'];

describe('Select component', () => {
  afterEach(() => {
    cleanup();
  });

  it('should render snapshot successfuly', () => {
    const props = {
      label: 'Test select',
      name: 'test-select',
      options: selectOptions,
      onChange: jest.fn(),
    };
    const {asFragment} = render(<Select {...props} />);

    expect(asFragment()).toMatchSnapshot();
  });

  it('should render all options inside when select has options', () => {
    const props = {
      label: 'Test select',
      name: 'test-select',
      options: selectOptions,
      onChange: jest.fn(),
    };

    render(<Select {...props} />);

    const options = screen.getAllByRole('option');

    expect(options).toHaveLength(selectOptions.length);
  });
});
