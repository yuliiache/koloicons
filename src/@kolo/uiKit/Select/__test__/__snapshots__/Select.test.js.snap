// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select component should render snapshot successfuly 1`] = `
<DocumentFragment>
  <div
    class="select"
  >
    <label
      class="selectLabel"
      data-testid="selectComponent"
      for="selectId"
    >
      Test select
    </label>
    <div
      class="selectWrapper"
    >
      <select
        class="selectOptions"
        id="selectId"
        name="test-select"
      >
        <option
          class="selectOption"
          data-testid="first"
          value="first"
        >
          first
        </option>
        <option
          class="selectOption"
          data-testid="second"
          value="second"
        >
          second
        </option>
        <option
          class="selectOption"
          data-testid="third"
          value="third"
        >
          third
        </option>
      </select>
    </div>
  </div>
</DocumentFragment>
`;
